### PR TITLE
Make BlockPayload::encode infallible

### LINKED ISSUE #3020: [Cleanup] - Remove `BlockPayload::encode` in favor of `EncodeBytes` bound
### What is this task and why do we need to work on it?

`EncodeBytes` is meant as a general trait for all data structures that need to be canonicaly encoded.

### What work will need to be done to complete this task?

`BlockPayload::encode` is currently fallible, point of failure being u32->usize conversion of transaction length. This check should be pushed to the point of transaction creation, since non-encodable transactions are basically useless.

### Are there any other details to include?

_No response_

### What are the acceptance criteria to close this issue?

`BlockPayload::encode` is removed, `BlockPayload: EncodeBytes` bound is added. 

### Branch work will be merged to (if not the default branch)

_No response_

### PR DESCRIPTION
**Pull request by QuentinI**
Closes #3020 

### This PR: 
- Removes `BlockPayload::encode`
- Adds `BlockPayload: EncodeBytes` bound instead
- Makes `TestTransaction::encode` infallible
- Makes un-encodable `TestTransaction` un-constructable

### Key places to review: 
`crates/example-types/src/block_types.rs`

### CODE REVIEW

File: crates/example-types/src/block_types.rs [hunk 1]
```diff
@@ -17,38 +17,77 @@ use hotshot_types::{
...
+    }
+}
 
 impl TestTransaction {
+    /// Construct new transaction
+    ///
+    /// # Panics
+    /// If `bytes.len()` > `u32::MAX`
+    pub fn new(bytes: Vec<u8>) -> Self {
+        Self::try_new(bytes).expect("Vector too long")
```
Comment (QuentinI) [inline crates/example-types/src/block_types.rs:47]: This `expect` may be controversial. My reasoning was
- 4Gb transaction is an absurd notion we shouldn't ever encounter
- Rust stdlib is allowed to panic on allocation failure, this isn't that different of a case
- this is a testing type anyway, no production code will ever call this function

Comment (jparr721) [reply crates/example-types/src/block_types.rs:47]: That's a reasonable justification.

File: crates/examples/infra/mod.rs [hunk 2]
```diff
@@ -472,7 +472,7 @@ pub trait RunDA<
                                         .get_transactions(leaf.get_block_header().metadata())
                                     {
                                         let restored_timestamp_vec =
-                                            tx.0[tx.0.len() - 8..].to_vec();
+                                            tx.bytes()[tx.bytes().len() - 8..].to_vec();
```
Comment (jparr721) [inline crates/examples/infra/mod.rs:475]: This is so much better. 10/10 change.

File: crates/example-types/src/block_types.rs [hunk 3]
```diff
@@ -17,38 +17,77 @@ use hotshot_types::{
...
+        Self::try_new(value).ok_or(TransactionTooLong)
+    }
+}
 
 impl TestTransaction {
+    /// Construct new transaction
+    ///
+    /// # Panics
+    /// If `bytes.len()` > `u32::MAX`
+    pub fn new(bytes: Vec<u8>) -> Self {
```
Comment (jparr721) [inline crates/example-types/src/block_types.rs:46]: Would this be better served as `impl From<Vec<u8>>` or something? That way we could go from bytes to tx and back? Open to your thoughts.

**Review by jparr721 — COMMENTED** [review]
Great work. I especially like the improved self-documentation of the `bytes` methods.

File: crates/example-types/src/block_types.rs (hunk 3 above)
Comment (QuentinI) [reply crates/example-types/src/block_types.rs:46]: It already implements `TryFrom<Vec<u8>>`. Do you mean ditching constructor entirely in favor of that implementation?

Comment (jparr721) [reply crates/example-types/src/block_types.rs:46]: Yeah that's what I was thinking. But if you prefer to have both, I'm not opposed. Just trying to think of it from a rust-y standpoint.

Comment (QuentinI) [reply crates/example-types/src/block_types.rs:46]: I think still having a constructor would be more rust-y, see for example all the wrapper types (`Cell`, `Arc`, `Mutex`, etc) in stdlib: they have corresponding `From` implementations, but provide `::new` as well.
Also, constructor is something that you almost definitely will want to call on a type, and to call `TestTransaction::try_from(x)` you'd have import `TryFrom` trait, which is a hurdle.

**Review by jparr721 — APPROVED** [review]
Sorry! I thought I approved this.

**Comment by QuentinI** [comment]
@jparr721 please re-approve, resolved conflicts

**Review by jparr721 — APPROVED** [review]

**Pull request merged**